### ISSUE TITLE
Feature-interp

### ISSUE THREAD

**Issue by georgebuzzell** [NDCLab/baseEEG#29]
Feature-interp
-interpolate missing channels, at the channel/epoch level using a spherical spline interpolation, as implemented in mne
-interpolate missing channels, at the global level, using a spherical spline interpolation, as implemented in mne
-write to annotations file which channels were interpolated and using what method

**Comment by georgebuzzell** [comment]
@Jonhas Thank you for self-assigning here! Please do not hesitate to reach out with any questions, etc, or if you want to jump on a call to discuss!

**Comment by Jonhas** [comment]
@georgebuzzell I have some trouble storing Polhemus digitization data. Currently, I have the following script. 
```python
#!/usr/bin/python3

import matplotlib
from pathlib import Path
import mne
import mne_bids
matplotlib.use('Qt5Agg')

# Read BIDS Data
bids_root = Path('eeg_matchingpennies')
bids_path = mne_bids.BIDSPath(subject='05',
                              task='matchingpennies',
                              datatype='eeg',
                              root=bids_root)
raw = mne_bids.read_raw_bids(bids_path)

bad_channels = raw.info['bads']
picks = mne.pick_channels_regexp(raw.ch_names, regexp='FC.')
```
where we will get the following bad channels.
` ['FC5', 'FC6'] ` 
I then do the following to pl
```python
picks = mne.pick_channels_regexp(raw.ch_names,regexp='FC.')
raw.plot(order=picks,n_channels=len(picks))
```
I then plot the channels, and see that the ones marked as "bad" are plotted in a light gray color in a layer 
gray color in a layer behind the other channels, to make it easy to distinguish them from 
"good" channels. This will produce 
![image](https://user-images.githubusercontent.com/45021859/114648440-68b20980-9ca4-11eb-912b-8dfb7160e744.png)
I then load our data into ram. 
```python 
raw.load_data()
```

I get an error shortly after, and this is from running the interpolate_bads function
```python
raw_data_copy = raw.copy().pick_types(meg=True,eeg=True,exclude=[])
print(raw.info['dig'])
raw_interp = raw_data_copy.copy().interpolate_bads(reset_bads=False)
```
From printing raw.info['dig'], we see that the dictionary of Polhemus digitization data in head coordinates is null. 
From mne's library, the runtime error is thrown in the following: 

```python
 if info['dig'] is None:
        raise RuntimeError('Cannot fit headshape without digitization '
                           ', info["dig"] is None')
```
I assume we need to import MEG data for this, is this correct? I am a liittle confused on how to approach this issue.

**Comment by georgebuzzell** [comment]
@Jonhas We actually do not want to import/use meg data. I believe this is an EEG-only dataset (at least, I hope so!). In this line, can you try setting meg=False:

raw_data_copy = raw.copy().pick_types(meg=True,eeg=True,exclude=[])

You should not require polhemus data to interpolate. You could use that in interpolation, for sure, but you should not need it. You should be able to interpolate using spatial coordinates provided by the montage file alone. 

The MEG and polhemus data are two different kinds of things. Happy to explain that further on a call? Can you try re-running with MEG=false and follow up? If you have a similar error still, then we need to figure out how to tell mne to look to the montage data for the coordinates (or to read that in properly to mne). Does that make sense?

**Comment by Jonhas** [comment]
@georgebuzzell Thank you for reply. That was what I did initially, since it is an EEG-only dateset after all, but it gave me an error, as shown on my previous comment. I read somewhere the mne documentation that CTF polhemus data can be pulled from MEG files, but It seems I definitely read that wrong. This is why I believed we should pick MEG data as well, (as a result have meg=True), but this did not solve my issue. If possible, could we also discuss this in the upcoming meeting? I think it would benefit the others to found out more on this datatype.

**Comment by Jonhas** [comment]
Now that I'm thinking about it, the reason I could of gotten an error is from the issue #38, i remember this error popping up as well at runtime, and I assumed it was mne conflicting with matplotlib. However, after reading the issue, that was not the case. I'll make sure to update my docker file.

**Comment by georgebuzzell** [comment]
@Jonhas glad to see you are making progress here! As an initial test to get things working, that was great thinking to just use a 1020 default montage. I believe that your approach of using  "mne.channels.make_standard montage" is essentially correct, though I will check the raw data for this example participant regarding where the montage info is stored. It SHOULD be stored somewhere with the raw data. In fact, we are currently discussing here https://github.com/NDCLab/baseEEG/issues/18 as to how/where we should store the montage info during feature-io.

For some additional background and context on standard channel locations vs digitizations, please see our discussion here: https://github.com/NDCLab/baseEEG/issues/18

Regarding MEG, you will not need to pull anything in from there (I think you realize this now). So no need to worry about using MEG in any way for a solution. However, I am, of course, more than happy to explain more about MEG and the difference between MEG and EEG at our next meeting!

TLDR: You seem to be TOTALLY on the right track. I will check the montage for these data and follow up

**Comment by georgebuzzell** [comment]
@jessb0t OK! I found where you can pull that info for the correct montage. If you look in the following bids meta data file:
![image](https://user-images.githubusercontent.com/71228105/114867135-6ba31c00-9dc2-11eb-8931-659fa5dd685d.png)

You will see that there is a field line that reads:

"CapManufacturersModelName": "actiCAP 64Ch Standard-2"

This line will tell you what cap was used for the EEG data, and then we just need to pull the appropriate montage file based on what is written here. standard_1020 as you used, will actually be appropriate here and is appropriate for the majority of caps. Ultimately, we would probably want code to read this line from the BIDS metadata file, and then to create the appropriate montage based on what was written here. For most caps, we would just default to standard_1020, but for some specific caps that diverge from standard_1020, we would pull in the appropriate montage file. 

I can see how this might be confusing, and I am not sure if I am fully explaining things well. Basically, "1020" is a standardized set of locations that many different EEG cap manufacturers conform to. So, for most caps, we can just use these standard locations. However, some manufacturers diverge, in which case we need to use non-standard electrode names or locations. Please let me know if this makes sense or not. Moreover, please let me know if you have additional questions before you can move forward. 

PS: Right now, you are writing code to lookup and set the montage file appropriately. This code is necessary for the feature you are working on, but it might also be needed for some (but definitely not all) the other features being worked on. Long story short is that the code you write here will ultimately go near the beginning of the pipeline, even though the actual interpolation itself will not happen until the end of the pipeline. However, for now, we are working on features independently, so do not worry about that for now. However, if someone else hits a similar issue, we will want to share the set montage code.

Thanks again, and great work!!

Tagging others here, as this set montage issue may be relevant for other features. @yanbin-niu @F-said @DMRoberts @SDOsmany

**Comment by DavyNeat** [comment]
@Jonhas Great job setting up the interpolation for the channels! I've just added a bit of code at the end of what you have done that would make an annotation of the channels that were interpolated.
 ``` python

 # Add annotations for the interpolated channels
interp_annot = mne.Annotations(onset=[0],
                              duration=[1],
                              description=['Interpolated with spherical spline'],
                              ch_names = [bad_channels])
raw_interp = raw_interp.copy().set_annotations(interp_annot)

 ```

@georgebuzzell  I'm assuming that once we have the annotations json completed that this information would be going there instead correct? Also, let me know if I'm missing any information for the annotations or if they should be set in some other location in the data.

**Comment by georgebuzzell** [comment]
@DavyNeat Great to see you on here contributing!

It looks like I may have caused some confusion by using the term for our output file. In particular, my understanding is that mne uses "annotations" to essentially refer to details associated with a particular point in time in the eeg file. In EEGLAB terms, we would call this a "marker" or an "event". My understanding is that mne does not use "annotations" to describe things not tied to a particular time point, for example, to describe what filter settings were used, or which channels were interpolated.

Thus, I don't think that you would want to modify mne.Annotations as you have done here. My understanding is that would not really be proper use of the mne.Annotations. However, I think you are absolutely on the right track, and again, I think the confusion is my fault. To move forward, I have changes the name of the output file from "annotations_preproc.Json" to "output_preproc.json" (see updated roadmap). You are absoulately correct that you would want to add this information there. Can you please take a stab at doing that?

@Jonhas @F-said @yanbin-niu @DMRoberts @SDOsmany

**Comment by georgebuzzell** [comment]
@Jonhas this looks really great! I love how clean your documentation is!

A minor comment: It looks like your function is set to take output_dict as an input, correct? We actually do not want these functions to take output_dict as an input. Instead, they should take user_params_dict as an input, and, they should create an output_dict (e.g. output_dict_interp, output_dict_filter, etc). Crucially, each function should create its own output_dict that is independent of the other functions. A final function in main will then collate all of these output_dict and write to json. Does that make sense?

@F-said @DMRoberts please comment as well

But again, @Jonhas the function as a whole looks fantastic!

**Comment by F-said** [comment]
Closing per merged PR #96
